networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/compression/SnappyCompressor.java

```diff
@@ -45,31 +47,60 @@ public Bytes compress(final Bytes data) {
     }
   }
 
+  @Override
+  public Bytes uncompress(final Bytes data) throws CompressionException {
+    try (final InputStream byteStream = new ByteArrayInputStream(data.toArrayUnsafe())) {
+      // Read everything
+      return uncompress(byteStream, OptionalInt.empty());
+    } catch (IOException e) {
+      throw new RuntimeException(
+          "Unexpected error encountered while preparing to uncompress bytes", e);
+    }
+  }
+
   @Override
   public Bytes uncompress(final InputStream input, final int uncompressedPayloadSize)
       throws CompressionException {
+    return uncompress(input, OptionalInt.of(uncompressedPayloadSize));
+  }
+
+  @Override
+  public int getMaxCompressedLength(final int uncompressedLength) {
+    // Return worst-case compression size
+    // See:
+    // https://github.com/google/snappy/blob/537f4ad6240e586970fe554614542e9717df7902/snappy.cc#L98
+    return 32 + uncompressedLength + uncompressedLength / 6;
+  }
+
+  private Bytes uncompress(final InputStream input, OptionalInt uncompressedPayloadSize)
+      throws CompressionException {
     // This is a bit of a hack - but we don't want to close the underlying stream when
     // we close the SnappyFramedInputStream
-    final UncloseableInputStream unclosableStream = new UncloseableInputStream(input);
-    // Limit the max number of bytes we're allowed to read
-    final InputStream limitedStream =
-        ByteStreams.limit(unclosableStream, getMaxCompressedLength(uncompressedPayloadSize));
+    InputStream srcStream = new UncloseableInputStream(input);
 
-    try (final InputStream snappyIn = new SnappyFramedInputStream(limitedStream)) {
-      final Bytes uncompressed = Bytes.wrap(snappyIn.readNBytes(uncompressedPayloadSize));
+    if (uncompressedPayloadSize.isPresent()) {
+      // Limit the max number of bytes we're allowed to read if we know the payload size
+      final int maxCompressedBytes = getMaxCompressedLength(uncompressedPayloadSize.getAsInt());
+      srcStream = ByteStreams.limit(srcStream, maxCompressedBytes);
+    }
+
+    final int maxBytesToRead = uncompressedPayloadSize.orElse(Integer.MAX_VALUE);
+    try (final InputStream snappyIn = new SnappyFramedInputStream(srcStream)) {
+      final Bytes uncompressed = Bytes.wrap(snappyIn.readNBytes(maxBytesToRead));
 
       // Validate payload is of expected size
-      if (uncompressed.size() < uncompressedPayloadSize) {
+      final boolean validateSize = uncompressedPayloadSize.isPresent();
+      if (validateSize && uncompressed.size() < uncompressedPayloadSize.getAsInt()) {
         throw new PayloadSmallerThanExpectedException(
             String.format(
                 "Expected %d bytes but only uncompressed %d bytes",
-                uncompressedPayloadSize, uncompressed.size()));
+                uncompressedPayloadSize.getAsInt(), uncompressed.size()));
       }
-      if (snappyIn.available() > 0) {
+      if (validateSize && snappyIn.available() > 0) {
```

Comment:
Shouldn't we be doing this check in all cases? ie even if we don't know what the expected uncompressed size is, we should check that we actually read all bytes from the gossip message.

Reply:
It turns out we're supposed to use different compression strategies for gossip vs rpc requests (snappy block-format vs framed-format compression).  So, reverted these changes.